runners/spark/3/src/main/java/org/apache/beam/runners/spark/structuredstreaming/translation/SparkSessionFactory.java

```diff
@@ -95,24 +122,37 @@ public static SparkSession getOrCreateSession(SparkStructuredStreamingPipelineOp
     if (options.getUseActiveSparkSession()) {
       return SparkSession.active();
     }
-    return sessionBuilder(options.getSparkMaster(), options.getAppName(), options.getFilesToStage())
-        .getOrCreate();
+    return sessionBuilder(options.getSparkMaster(), options).getOrCreate();
   }
 
   /** Creates Spark session builder with some optimizations for local mode, e.g. in tests. */
   public static SparkSession.Builder sessionBuilder(String master) {
-    return sessionBuilder(master, null, null);
+    return sessionBuilder(master, null);
   }
 
   private static SparkSession.Builder sessionBuilder(
-      String master, @Nullable String appName, @Nullable List<String> jars) {
-    SparkConf sparkConf = new SparkConf();
-    sparkConf.setMaster(master);
-    if (appName != null) {
-      sparkConf.setAppName(appName);
-    }
-    if (jars != null && !jars.isEmpty()) {
-      sparkConf.setJars(jars.toArray(new String[0]));
+      String master, @Nullable SparkStructuredStreamingPipelineOptions options) {
+
+    SparkConf sparkConf = new SparkConf().setIfMissing("spark.master", master);
+    master = sparkConf.get("spark.master"); // update to effective master
+
+    if (options != null) {
+      sparkConf.setAppName(options.getAppName());
+
+      if (options.getFilesToStage() != null && !options.getFilesToStage().isEmpty()) {
+        // Append the files to stage provided by the user to `spark.jars`.
+        PipelineResources.prepareFilesForStaging(options);
+        String[] staged = filesToStage(options, Collections.emptyList());
+        String[] jars = sparkJars(sparkConf);
+        sparkConf.setJars(jars.length > 0 ? ArrayUtils.addAll(jars, staged) : staged);
+      } else if (!sparkConf.contains("spark.jars") && !master.startsWith("local[")) {
+        // Stage classpath if `spark.jars` not set and not in local mode.
+        PipelineResources.prepareFilesForStaging(options);
```

Comment:
`prepareFilesForStaging()` will do nothing as in that case `options.getFilesToStage()` is either null or empty. It will de equivalent to calling options.setFilesToStage() with an empty list

Reply:
It's the opposite @echauchot , this is actually the interesting case where it stages the entire classpath.
```Java
public static void prepareFilesForStaging(FileStagingOptions options) {
    List<String> filesToStage = options.getFilesToStage();
    if (filesToStage == null || filesToStage.isEmpty()) {
      filesToStage = detectClassPathResourcesToStage(ReflectHelpers.findClassLoader(), options);
      LOG.info(
          "PipelineOptions.filesToStage was not specified. "
              + "Defaulting to files from the classpath: will stage {} files. "
              + "Enable logging at DEBUG level to see which files will be staged.",
          filesToStage.size());
      LOG.debug("Classpath elements: {}", filesToStage);
    }
    final String tmpJarLocation =
        MoreObjects.firstNonNull(options.getTempLocation(), System.getProperty("java.io.tmpdir"));
    final List<String> resourcesToStage = prepareFilesForStaging(filesToStage, tmpJarLocation);
    options.setFilesToStage(resourcesToStage);
  }
  ```

Reply:
But unexpected, I agree!

Reply:
oops, I missed the `filesToStage = detectClassPathResourcesToStage(ReflectHelpers.findClassLoader(), options);` call. Nevermind !